packages/core/src/bookmark-manager.ts

```diff
@@ -0,0 +1,166 @@
+   */
+  forget (databases: string[]): void {
+    throw new Error('Not implemented')
+  }
+}
+
+export interface BookmarkManagerConfig {
+  initialBookmarks?: Map<string, string[]>
+  bookmarksSupplier?: (database?: string) => string[]
+  bookmarksConsumer?: (database: string, bookmarks: string[]) => void
+}
+
+/**
+ * @typedef {Object} BookmarkManagerConfig
+ * @property {Map<string,string[]>} [initialBookmarks] Defines the initial set of bookmarks. The key is the database name and the values are the bookmarks.
+ * @property {function([database]: string):string[]} [bookmarksSupplier] Called for supplying extra bookmarks to the BookmarkManager
+ * 1. supplying bookmarks from the given database when the default BookmarkManager's `.getBookmarks(database)` gets called.
+ * 2. supplying all the bookmarks when the default BookmarkManager's  `.getAllBookmarks()` gets called
+ * @property {function(database: string, bookmarks: string[]): void} [bookmarksConsumer] Called when the set of bookmarks for database get updated
+ */
+/**
+ * Provides an configured {@link BookmarkManager} instance.
+ *
+ * @param {BookmarkManagerConfig} [config={}]
+ * @returns {BookmarkManager}
+ */
+export function bookmarkManager (config: BookmarkManagerConfig = {}): BookmarkManager {
+  const initialBookmarks = new Map<string, Set<string>>()
+
+  config.initialBookmarks?.forEach((v, k) => initialBookmarks.set(k, new Set(v)))
+
+  return new Neo4jBookmarkManager(
+    initialBookmarks,
+    config.bookmarksSupplier,
+    config.bookmarksConsumer
+  )
+}
+
+class Neo4jBookmarkManager implements BookmarkManager {
+  constructor (
+    private readonly _bookmarksPerDb: Map<string, Set<string>>,
+    private readonly _bookmarksSupplier?: (database?: string) => string[],
+    private readonly _bookmarksConsumer?: (database: string, bookmark: string[]) => void
+  ) {
+
+  }
+
+  updateBookmarks (database: string, previousBookmarks: string[], newBookmarks: string[]): void {
+    const bookmarks = this._getOrInitializeBookmarks(database)
+    previousBookmarks.forEach(bm => bookmarks.delete(bm))
+    newBookmarks.forEach(bm => bookmarks.add(bm))
+
+    if (typeof this._bookmarksConsumer === 'function') {
+      this._bookmarksConsumer(database, [...bookmarks])
+    }
+  }
+
+  private _getOrInitializeBookmarks (database: string): Set<string> {
+    let maybeBookmarks = this._bookmarksPerDb.get(database)
+    if (maybeBookmarks == null) {
+      maybeBookmarks = new Set()
+      this._bookmarksPerDb.set(database, maybeBookmarks)
+    }
+    return maybeBookmarks
+  }
+
+  getBookmarks (database: string): string[] {
+    const bookmarks = this._bookmarksPerDb.get(database) ?? []
+
+    if (typeof this._bookmarksSupplier === 'function') {
+      const suppliedBookmarks = this._bookmarksSupplier(database) ?? []
+      return [...bookmarks, ...suppliedBookmarks]
+    }
+
+    return [...bookmarks]
+  }
+
+  getAllBookmarks (): string[] {
+    const bookmarks = []
```

Comment:
Yes, I was thinking about it during the implementation. The methods of `Set` are not that ergonomic as the ones from `Array` in JS and the value will transformed to `Bookmarks`, which removes duplicates. However, for an external user, it could a bit of pain have duplicates in this list.